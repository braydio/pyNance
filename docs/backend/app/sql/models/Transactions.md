# ðŸ“˜ `Transaction` Model

```markdown
# Transaction Model

## Purpose

Stores individual cash-flow events for a user account. Rows originate from Plaid
syncs or manual entries and power downstream balance calculations, budgeting, and
forecasting features.

## Fields

| Column                               | Type                          | Notes                                                                         |
| ------------------------------------ | ----------------------------- | ----------------------------------------------------------------------------- |
| `id`                                 | Integer (PK)                  | Auto-incrementing identifier used internally.                                 |
| `transaction_id`                     | String(64), unique, indexed   | Primary business key from Plaid or manual import; enforced unique constraint. |
| `user_id`                            | String(64), indexed, nullable | Owning user identifier for multi-tenant filtering.                            |
| `account_id`                         | String(64) FK                 | References `Account.account_id`; cascades on account deletion.                |
| `amount`                             | Numeric(18, 2)                | Stored as `Decimal`; preserves provider sign for inflow/outflow logic.        |
| `date`                               | DateTime(timezone=True)       | Timestamp of when the transaction occurred.                                   |
| `description`                        | String(256)                   | Raw provider description or user-supplied label.                              |
| `provider`                           | Enum[`manual`, `plaid`]       | Normalized via validator; defaults to `manual`.                               |
| `merchant_name`                      | String(128)                   | Resolved merchant display name (defaults to `"Unknown"`).                     |
| `merchant_type`                      | String(64)                    | Provider merchant type/category (defaults to `"Unknown"`).                    |
| `user_modified`                      | Boolean                       | Flagged when a human edits the transaction.                                   |
| `user_modified_fields`               | Text                          | Comma-separated/JSON list of manually edited attributes.                      |
| `updated_by_rule`                    | Boolean                       | Indicates a `TransactionRule` mutation applied during sync.                   |
| `category_id`                        | Integer FK, nullable          | Optional FK to `Category.id`; nullified on category deletion.                 |
| `category`                           | String(128)                   | Denormalized category label maintained beside `category_id`.                  |
| `personal_finance_category`          | JSON                          | Raw Plaid personal finance category payload.                                  |
| `personal_finance_category_icon_url` | String                        | Icon URL associated with Plaid personal finance category.                     |
| `pending`                            | Boolean                       | True while Plaid reports the entry as unsettled.                              |
| `is_internal`                        | Boolean, indexed              | Marks internal transfers for reporting exclusions.                            |
| `internal_match_id`                  | String(64), nullable          | Stores counterpart `transaction_id` when `is_internal` is set.                |

## Relationships

- `plaid_meta`: One-to-one with `PlaidTransactionMeta` (cascade delete-orphan) for
  detailed Plaid metadata.
- `recurrence_rule`: Backref from `RecurringTransaction` when this row seeds a recurring
  pattern.

## Behaviors

- Plaid transaction syncs upsert the `plaid_meta` relationship via
  [`refresh_or_insert_plaid_metadata`](../../../../../backend/app/sql/refresh_metadata.py),
  ensuring Plaid metadata stays aligned with the base row.
- Internal transfer detection in
  [`account_logic.detect_internal_transfer`](../../../../../backend/app/sql/account_logic.py)
  flags matching debits/credits, toggling `is_internal` and wiring
  `internal_match_id` pairs to prevent double-counting balances.
- The provider validator keeps `provider` constrained to `manual`/`plaid`, defaulting
  to `manual` when unknown values arrive.

## Related Logic

- [`transaction_models.py`](../../../../../backend/app/models/transaction_models.py)
- [`transactions_logic.md`](../transactions_logic.md)
- [`recurring_logic.md`](../recurring_logic.md)

## Related Docs

- [`transactions.md`](../../services/transactions.md)
- [`enhanced_account_history.md`](../../services/enhanced_account_history.md)
- [`sync_service.md`](../../services/sync_service.md)
```
