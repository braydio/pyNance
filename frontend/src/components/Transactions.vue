<template>
    <div>
      <h1>Transactions</h1>
      <table>
        <thead>
          <tr>
            <th>Date</th>
            <th>Amount</th>
            <th>Name</th>
            <th>Category</th>
          </tr>
        </thead>
        <tbody>
          <tr v-for="tx in transactions" :key="tx.id">
            <td>{{ tx.date }}</td>
            <td>{{ tx.amount }}</td>
            <td>{{ tx.name }}</td>
            <td>{{ tx.category }}</td>
          </tr>
        </tbody>
      </table>
    </div>
  </template>
  
  <script>
  import axios from "axios";
  
  export default {
    name: "Transactions",
    data() {
      return {
        transactions: [],
      };
    },
    async created() {
      await this.fetchTransactions();
    },
    methods: {
      async fetchTransactions() {
        try {
          const response = await axios.get("/get_transactions");
          this.transactions = response.data.data.transactions;
        } catch (error) {
          console.error("Failed to fetch transactions:", error);
        }
      },
    },
  };
  </script>
  