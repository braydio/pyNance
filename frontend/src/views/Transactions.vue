
<template>
  <div class="transactions-page">
    <header class="transactions-header">
      <div>
        <h1 class="heading-lg">Transactions</h1>
        <h3 class="text-muted mb-2">View and manage your transactions</h3>
              </div>
    </header>

    <main>
      <div class="controls mb-2">
        <input
          v-model="searchQuery"
          type="text"
          placeholder="Search transactions..."
          class="search-input"
        />
      </div>

      <TransactionsTable
        :transactions="filteredTransactions"
        :sort-key="sortKey"
        :sort-order="sortOrder"
        @sort="setSort"
      />

      <div id="pagination-controls" class="mt-3 flex-center gap-2">
        <button class="btn" @click="changePage(-1)" :disabled="currentPage === 1">
          Previous
        </button>
        <span class="text-muted">Page {{ currentPage }} of {{ totalPages }}</span>
        <button class="btn" @click="changePage(1)" :disabled="currentPage >= totalPages">
          Next
        </button>
        <div>
          <RecurringTransactionSection provider="plaid" />
        </div>
      </div>
    </main>
  </div>
</template>

<script>
import { useTransactions } from "@/composables/useTransactions.js";
import TransactionsTable from "../components/UpdateTransactionsTable.vue";
import RecurringTransactionSection from "../components/RecurringTransactionSection.vue";

export default {
  name: "Transactions",
  components: { TransactionsTable, RecurringTransactionSection },
  setup() {
    const {
      searchQuery,
      currentPage,
      totalPages,
      filteredTransactions,
      changePage,
      sortKey,
      sortOrder,
      setSort,
    } = useTransactions(15);

    return {
      searchQuery,
      currentPage,
      totalPages,
      filteredTransactions,
      changePage,
      sortKey,
      sortOrder,
      setSort,
    };
  },
};
</script>

<style scoped>
.transactions-page {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  background-color: var(--page-bg);
  color: var(--color-text-light);
  padding: 1rem;
  gap: 2rem;
}
.transactions-header {
  background-color: var(--color-bg-secondary);
  padding: 1.5rem 1rem;
  text-align: center;
  border-bottom: 1px solid var(--themed-border);
  border-radius: 12px;
  box-shadow: 0 4px 12px var(--shadow);
}
.transactions-header h1 {
  margin: 0;
  color: var(--color-accent-yellow);
  font-size: 2rem;
  text-shadow: 0 0 6px var(--);
}

.controls {
  margin-bottom: 1rem;
}

.search-input {
  padding: 0.5rem 1rem;
  width: 100%;
  max-width: 320px;
  border: 1px solid var(--divider);
  border-radius: 6px;
  background-color: var(--color-bg-secondary);
  color: var(--color-text-light);
  font-family: var(--font-mono);
}

#pagination-controls {
  margin-top: 1rem;
  text-align: center;
}

button.btn {
  background-color: var(--button-bg);
  color: var(--color-text-light);
  border: 1px solid var(--divider);
  border-radius: 6px;
  padding: 0.5rem 1rem;
  cursor: pointer;
  transition: background-color 0.2s ease, transform 0.2s ease;
}

button.btn:hover:not(:disabled) {
  background-color: var(--button-hover-bg);
  transform: translateY(-1px);
  box-shadow: 0 0 6px var(--hover-glow);
}

button.btn:disabled {
  background-color: #444;
  opacity: 0.5;
  cursor: not-allowed;
}
</style>

