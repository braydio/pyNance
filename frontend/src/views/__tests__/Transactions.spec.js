// @vitest-environment jsdom
import { describe, it, expect, vi } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import Transactions from '../Transactions.vue'

vi.mock('@/api/transactions', () => ({
  fetchTransactions: vi.fn().mockResolvedValue({ transactions: [], total: 0 })
}))

vi.mock('vue-router', () => ({
  useRoute: () => ({ query: {} })
}))

describe('Transactions.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(Transactions, {
      global: {
        stubs: ['ImportFileSelector', 'UpdateTransactionsTable', 'RecurringTransactionSection', 'Button', 'Card', 'CreditCard', 'BasePageLayout']
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
