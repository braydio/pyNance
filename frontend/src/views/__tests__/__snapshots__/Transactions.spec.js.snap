// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transactions.vue > matches snapshot 1`] = `
"<div class="transactions-page container space-y-8">
  <!-- Header -->
  <card-stub class="p-6 flex items-center gap-3"></card-stub><!-- Top Controls -->
  <card-stub class="p-6"></card-stub><!-- Main Table -->
  <card-stub class="p-6 space-y-4"></card-stub><!-- Pagination -->
  <div id="pagination-controls" class="flex items-center justify-center gap-4">
    <button-stub variant="outline" pill="false" disabled="true"></button-stub><span class="text-muted">Page 1 of 1</span>
    <button-stub variant="primary" pill="false" disabled="true"></button-stub>
  </div><!-- Recurring Transactions -->
  <card-stub class="p-6 space-y-4"></card-stub>
</div>"
`;
